### ISSUE TITLE
Adopt and enforce new `WP_CLI_CS` standard

### ISSUE THREAD

**Issue by schlessera** [wp-cli/wp-cli#5166]
We have a new PHPCS standard for WP-CLI called `WPCliCS` (props @jrfnl). It is part of the `wp-cli/wp-cli-tests` package starting with version v2.1.0.

To adopt & enforce this new standard, the following actions need to be taken for this repository:

- [ ] Create a PR that adds a custom ruleset `phpcs.xml.dist` to the repository

    - Add `phpcs.xml.dist` file
    - Adapt `.distignore` to ignore `phpcs.xml.dist` & `phpunit.xml.dist`
    - Adapt `.gitignore` to ignore `phpunit.xml`, `phpcs.xml` & `.phpcs.xml`
    - Require version `^2.1` of the `wp-cli/wp-cli-tests` as a dev dependency
- [ ] Make any required changes to the code that fail the checks from the above ruleset in separate PRs
- [ ] Merge thre ruleset once all required changes have been processed and merged

A sample PR for a simple repository can be seen here: https://github.com/wp-cli/maintenance-mode-command/pull/3/files

Related https://github.com/wp-cli/wp-cli/issues/5179

**Comment by jrfnl** [comment]
WIP branch with the ruleset and currently added whitelist comments can be found here: https://github.com/jrfnl/wp-cli/tree/feature/update-phpcs

List of currently remaining issues which need to be addressed: 
[20190417-phpcs-result-wpcli.txt](https://github.com/wp-cli/wp-cli/files/3090950/20190417-phpcs-result-wpcli.txt)

**Comment by jrfnl** [comment]
FYI: WIP ruleset branch updated after the release of `wp-cli-tests` 2.1.0.

**Comment by jrfnl** [comment]
Updated remaining issue list: 
[20190419-phpcs-result-wpcli.txt](https://github.com/wp-cli/wp-cli/files/3098463/20190419-phpcs-result-wpcli.txt)

This list is based on `master` + PR #5171 + PR #5175

**Comment by schlessera** [comment]
My list of remaining issues after all PRs are merged:
```
// Whitelist rule for the entire file. The file itself is a "hack" already.
FILE: php\wp-settings-cli.php
------------------------------------------------------------------------------------------
 266 | ERROR | Global variables defined by a theme/plugin should start with the
     |       | theme/plugin prefix. Found: "$symlinked_plugins_supported".
     |       | (WordPress.NamingConventions.PrefixAllGlobals.NonPrefixedVariableFound)
 272 | ERROR | Global variables defined by a theme/plugin should start with the
     |       | theme/plugin prefix. Found: "$mu_plugin".
     |       | (WordPress.NamingConventions.PrefixAllGlobals.NonPrefixedVariableFound)
 279 | ERROR | Global variables defined by a theme/plugin should start with the
     |       | theme/plugin prefix. Found: "$network_plugin".
     |       | (WordPress.NamingConventions.PrefixAllGlobals.NonPrefixedVariableFound)
 312 | ERROR | Global variables defined by a theme/plugin should start with the
     |       | theme/plugin prefix. Found: "$plugin".
     |       | (WordPress.NamingConventions.PrefixAllGlobals.NonPrefixedVariableFound)
 394 | ERROR | Global variables defined by a theme/plugin should start with the
     |       | theme/plugin prefix. Found: "$locale_file".
     |       | (WordPress.NamingConventions.PrefixAllGlobals.NonPrefixedVariableFound)
 440 | ERROR | Global variables defined by a theme/plugin should start with the
     |       | theme/plugin prefix. Found: "$file".
     |       | (WordPress.NamingConventions.PrefixAllGlobals.NonPrefixedVariableFound)
------------------------------------------------------------------------------------------


// Not sure why that one is singled out. Whitelist, I suppose.
FILE: php\WP_CLI\Runner.php
------------------------------------------------------------------------------------------
 1246 | ERROR   | Global variables defined by a theme/plugin should start with the
      |         | theme/plugin prefix. Found: "$current_blog".
      |         | (WordPress.NamingConventions.PrefixAllGlobals.NonPrefixedVariableFound)
------------------------------------------------------------------------------------------

// File should be whitelist like we did with `behat-tags`, it is a stand-alone script.
FILE: utils\get-package-require-from-composer.php
------------------------------------------------------------------------------------------
  3 | ERROR | Global variables defined by a theme/plugin should start with the
    |       | theme/plugin prefix. Found: "$file".
    |       | (WordPress.NamingConventions.PrefixAllGlobals.NonPrefixedVariableFound)
  9 | ERROR | Global variables defined by a theme/plugin should start with the
    |       | theme/plugin prefix. Found: "$contents".
    |       | (WordPress.NamingConventions.PrefixAllGlobals.NonPrefixedVariableFound)
 10 | ERROR | Global variables defined by a theme/plugin should start with the
    |       | theme/plugin prefix. Found: "$composer".
    |       | (WordPress.NamingConventions.PrefixAllGlobals.NonPrefixedVariableFound)
------------------------------------------------------------------------------------------
```

**Comment by jrfnl** [comment]
@schlessera I've ran some tests and I still have a few more left (after excluding/whitelisting the above) and based on `master` + #5182 + #5183 + #5184 

```
FILE: php\utils-wp.php
------------------------------------------------------------------------------------------
FOUND 0 ERRORS AND 1 WARNING AFFECTING 1 LINE
------------------------------------------------------------------------------------------
 111 | WARNING | strip_tags() is discouraged. Use the more comprehensive
     |         | wp_strip_all_tags() instead.
     |         | (WordPress.WP.AlternativeFunctions.strip_tags_strip_tags)
------------------------------------------------------------------------------------------


FILE: php\utils.php
------------------------------------------------------------------------------------------
FOUND 2 ERRORS AFFECTING 2 LINES
------------------------------------------------------------------------------------------
 1033 | ERROR | [ ] Function name "isPiped" is not in snake case format, try
      |       |     "is_piped"
      |       |     (WordPress.NamingConventions.ValidFunctionName.FunctionNameInvalid)
------------------------------------------------------------------------------------------
PHPCBF CAN FIX THE 1 MARKED SNIFF VIOLATIONS AUTOMATICALLY
------------------------------------------------------------------------------------------


FILE: php\WP_CLI\Runner.php
------------------------------------------------------------------------------------------
FOUND 2 ERRORS AFFECTING 2 LINES
------------------------------------------------------------------------------------------
 1517 | ERROR | Hook names invoked by a theme/plugin should start with the theme/plugin
      |       | prefix. Found: "cli_init".
      |       | (WordPress.NamingConventions.PrefixAllGlobals.NonPrefixedHooknameFound)
------------------------------------------------------------------------------------------
```

Additionally, there are the following issues which don't need to be actioned:

This is a bug in PHPCS itself, a fix for which is already included in https://github.com/squizlabs/PHP_CodeSniffer/pull/2456, so should hopefully be part of PHPCS 3.5.0.
For now, we could just add a blank line below that statement or temporarily whitelist it with reference to the upstream PR.
```
FILE: php\utils.php
------------------------------------------------------------------------------------------
FOUND 2 ERRORS AFFECTING 2 LINES
------------------------------------------------------------------------------------------
  579 | ERROR | [x] There must be one blank line after the namespace declaration
      |       |     (PSR2.Namespaces.NamespaceDeclaration.BlankLineAfter)
------------------------------------------------------------------------------------------
PHPCBF CAN FIX THE 1 MARKED SNIFF VIOLATIONS AUTOMATICALLY
------------------------------------------------------------------------------------------
```


Once [your fix to WPCS](https://github.com/WordPress-Coding-Standards/WordPress-Coding-Standards/pull/1693) has been released, this error will be thrown - I have annotated this in the whitelist comment which will be added when the ruleset will be pulled (included in the WIP ruleset branch).
```
FILE: php\WP_CLI\Runner.php
------------------------------------------------------------------------------------------
FOUND 2 ERRORS AFFECTING 2 LINES
------------------------------------------------------------------------------------------
 1250 | ERROR | Overriding WordPress globals is prohibited. Found assignment to
      |       | $current_blog (WordPress.WP.GlobalVariablesOverride.Prohibited)
------------------------------------------------------------------------------------------
```

**Comment by jrfnl** [comment]
WIP ruleset branch has been rebased & updated: https://github.com/jrfnl/wp-cli/tree/feature/update-phpcs

**Comment by schlessera** [comment]
```
strip_tags() is discouraged. Use the more comprehensive wp_strip_all_tags() instead.
```
I'll have to look into that one, I think there was an issue before when I tried to use the new failsafe helper function.

```
Function name "isPiped" is not in snake case format, try "is_piped"
```
This should be whitelisted in the code, we cannot change that one.
Shall I do a separate PR, or will you add it to your branch?

```
Hook names invoked by a theme/plugin should start with the theme/plugin prefix. Found: "cli_init".
```
I'd add `cli` as prefix configuration for `WordPress.NamingConventions.PrefixAllGlobals.NonPrefixedHooknameFound` with a mention `Hooks.` in the ruleset.

**Comment by jrfnl** [comment]
> Shall I do a separate PR, or will you add it to your branch?

I'll add it to my branch.

> I'd add cli as prefix configuration for WordPress.NamingConventions.PrefixAllGlobals.NonPrefixedHooknameFound with a mention Hooks. in the ruleset.

In the `WP_CLI_CS` ruleset or in the project repo custom ruleset ?
Also: be aware that while we may document that a prefix is only supposed to be used for hooks, the sniffs does not distinguish between uses and will then allow it for any use-case.

Would it be an idea to deprecated the hook instead and replace it with a properly prefixed one ?
That would mean the code would change to something along the lines of:
```php
		WP_CLI::add_wp_hook(
			'init',
			function () {
				do_action_deprecated( 'cli_init', array(), 'WP_CLI v #.#, 'wpcli_init' );
				do_action( 'wpcli_init' );
			}
		);
```

**Comment by schlessera** [comment]
> In the WP_CLI_CS ruleset or in the project repo custom ruleset ?

In the project repo custom ruleset.

> Also: be aware that while we may document that a prefix is only supposed to be used for hooks, the sniffs does not distinguish between uses and will then allow it for any use-case.

Noted, but I don't think it's an issue if the prefix still is `cli` here.

> Would it be an idea to deprecated the hook instead and replace it with a properly prefixed one ?

No, I don't think that's necessary or of any benefit.

**Comment by jrfnl** [comment]
> This should be whitelisted in the code, we cannot change that one.

@schlessera Updating the branch now - what can I add a the reason, i.e. why can't it be changed ? BC or because an external dependency expects that name or ... ?

**Comment by schlessera** [comment]
Other packages depend on it, and it would be a big undertaking to change them all in sync.

**Comment by jrfnl** [comment]
OK, so BC basically.

Only the `strip_tags` issue remaining in that case.

**Comment by jrfnl** [comment]
Now PR #5182 has been merged, I have pulled the ruleset: #5198
